### ISSUE TITLE
[Feature]: How about change AsyncBoundaryProvider(@toss/async-boundary) to ResetErrorBoundaryProvider(@toss/error-boundary)?

### ISSUE THREAD

**Issue by manudeli** [toss/slash#144]
# How about change [AsyncBoundaryProvider(@toss/async-boundary)](https://github.com/toss/slash/blob/bec5420c15b8b3b6d28dfefac08d56b8bc46e0a0/packages/react/async-boundary/src/provider/AsyncBoundaryContext.tsx) to ResetErrorBoundaryProvider(@toss/error-boundary)?

### **Package Scope**
async-boundary, error-boundary

- [x] Add to an existing package

### **Overview**

Current AsyncBoundaryProvider is not scoped AsyncBoundary, I think it is just designed only for reseting resetKey ErrorBoundary

## AS-IS: AsyncBoundaryProvider`(@toss/async-boundary)`
```ts
interface AsyncBoundaryProvider {
  resetKey: number;
  reset(): void;
}

const AsyncBoundaryContext = createContext<AsyncBoundaryProvider | null>(null);

interface Props {
  children: ReactNode;
}

export function AsyncBoundaryProvider({ children }: Props) {
  const [resetKey, reset] = useResetError();

  return <AsyncBoundaryContext.Provider value={{ resetKey, reset }}>{children}</AsyncBoundaryContext.Provider>;
}

export function useAsyncBoundaryContext() {
  return useContext(AsyncBoundaryContext) ?? { resetKey: null, reset: () => {} };
}

```

### **Describe the solution you'd like**

If I can edit, I want to change AsyncBoundaryProvider to ResetErrorBoundaryProvider

## TO-BE: ResetErrorBoundaryProvider`(@toss/error-boundary)`
```ts
const ResetErrorBoundaryContext = createContext({
  resetKey: 0,
  reset: () => {},
});
if (process.env.NODE_ENV !== 'production') {
  ResetErrorBoundaryContext.displayName = 'ResetErrorBoundaryContext';
}

interface Props {
  children: ReactNode;
}

export function ResetErrorBoundaryProvider({ children }: Props) {
  const [resetKey, reset] = useResetError();

  return <ResetErrorBoundaryContext.Provider value={{ resetKey, reset }}>{children}</ResetErrorBoundaryContext.Provider>;
}

export function useResetErrorBoundaryContext() {
  return useContext(ResetErrorBoundaryContext);
}

```

#### AS-IS(`ReturnType<typeof useAsyncBoundaryContext>`)
union type is not intuitive.

```ts
import { expectType } from 'tsd';
expectType<AsyncBoundaryProvider | { resetKey: number; reset(): void; }>(useAsyncBoundaryContext()) // not intuitive, not essential union type 
```

<img width="804" alt="image" src="https://user-images.githubusercontent.com/61593290/201651513-2effe032-626b-4ca6-bc09-5a81222b9a18.png">

#### TO-BE(`ReturnType<typeof useResetErrorBoundaryContext>`)
infer just one type like below picture

```ts
import { expectType } from 'tsd';
expectType<{ resetKey: number; reset(): void; }>(useResetErrorBoundaryContext()) // intuitive, just essential type 
```
<img width="659" alt="image" src="https://user-images.githubusercontent.com/61593290/201652682-96cf3815-6256-4f3b-84ff-bab255783db1.png">

## 2. add displayName Context only on development mode
we can check displayName, `ResetErrorBoundaryContext.Provider` on react-devtools only for development mode

```ts
const ResetErrorBoundaryContext = createContext({
  resetKey: 0,
  reset: () => {},
});
if (process.env.NODE_ENV !== 'production') {
  ResetErrorBoundaryContext.displayName = 'ResetErrorBoundaryContext';
}

```

## 3. How about ResetErrorBoundaryProvider.children can use reset easily like this

If I can't use useContext, for example in Class Component or HTML like button or div, It will be helpful 

### ResetErrorBoundaryProvider
```ts
export function ResetErrorBoundaryProvider({ children: Children }: Props) {
  const [resetKey, reset] = useResetError();

  const node = typeof Children === 'function' ? <Children reset={reset} /> : Children;

  return <ResetErrorBoundaryContext.Provider value={{ resetKey, reset }}>{node}</ResetErrorBoundaryContext.Provider>;
}
```

### Use Case
```ts
<ResetErrorBoundaryProvider>
  {({ reset }) => (
	<button onClick={reset} /*  <-- error reset 해야 하는 부분 */ />
	<AsyncBoundary /* <-- suspense + error boundary로 감싸져야 하는 부분 */
	    pendingFallback={<TableSkeleton title="상세내역" row={10} />}
	    errorFallback={({ error, reset }) => (
	      <ErrorAlert theme="yellow" error={error} message="다시 시도해주세요." onResetError={reset} />
	    )}
	>
      <DataViewer />
    </AsyncBoundary>
  )}
</ResetErrorBoundaryProvider>
```

Can I change AsyncBoundaryProvider like this? this will be BREAKING CHANGE, so I want to check it before starting code

**Comment by HoseungJang** [comment]
About renaming `AsyncBoundaryProvider` to `ResetErrorBoundaryProvider`, I think the current naming is appropriate.

Because I think `AsyncBoundaryProvider` is not just for resetting errors and it provides controls for `AsyncBoundary`. Currently there is only the control to reset but I think any other controls could be added in the future.

---

> 3. How about ResetErrorBoundaryProvider.children can use reset easily like this

I think this pattern is far from the concept of context provider. I think it's the same as [consumer](https://ko.reactjs.org/docs/context.html#contextconsumer).

**Comment by HoseungJang** [comment]
About renaming `AsyncBoundaryProvider` to `ResetErrorBoundaryProvider`, I think the current naming is appropriate.

Because I think `AsyncBoundaryProvider` is not just for resetting errors, and it provides controls for `AsyncBoundary`. Currently there is only the control to reset errors but I think any other controls could be added in the future.

And the naming `AsyncBoundaryProvider` is more clear because developers can intuitively see what they exactly reset:

```typescript
const asyncBoundary = useAsyncBoundaryContext();
asyncBoundary.reset(); // resets only AsyncBoundary


const { reset } = useResetErrorBoundaryContext();
reset();
// resets error boundary but what and where is it?
```

---

> 3. How about ResetErrorBoundaryProvider.children can use reset easily like this

I think this pattern is far from the concept of context provider. I think it's the same as [consumer](https://ko.reactjs.org/docs/context.html#contextconsumer).

**Comment by manudeli** [comment]
> > 3. How about ResetErrorBoundaryProvider.children can use reset easily like this
> 
> I think this pattern is far from the concept of context provider. It looks like [context consumer](https://ko.reactjs.org/docs/context.html#contextconsumer).

I understood it. thanks! Provider is just Provider right!

**Comment by manudeli** [comment]
> About renaming `AsyncBoundaryProvider` to `ResetErrorBoundaryProvider`, I think the current naming is appropriate.
> 
> Because I think `AsyncBoundaryProvider` is not just for resetting errors, and it provides controls for `AsyncBoundary`. Currently there is only the control to reset errors but I think any other controls could be added in the future.
> 
> And the naming `AsyncBoundaryProvider` is more clear because developers can intuitively see what they exactly reset:
> 
> ```ts
> const asyncBoundary = useAsyncBoundaryContext();
> asyncBoundary.reset(); // resets only AsyncBoundary of `@toss/async-boundary`
> 
> 
> const { reset } = useResetErrorBoundaryContext();
> reset(); // resets error boundary but..
> ```

then, How about other way: AsyncBoundaryProvider(@toss/async-boundary) depends on ErrorBoundaryProvider(@toss/error-boundary). AsyncBoundaryProvider wrapping ErrorBoundaryProvider like [AsyncBoundary](https://github.com/toss/slash/blob/main/packages/react/async-boundary/src/AsyncBoundary.tsx) wrapping ErrorBoundary
```ts
/** @tossdocs-ignore */
import { ErrorBoundary } from '@toss/error-boundary';
import { ComponentProps, forwardRef, Ref, Suspense, useImperativeHandle, useRef } from 'react';

type ErrorBoundaryProps = Omit<ComponentProps<typeof ErrorBoundary>, 'renderFallback'>;
type SuspenseProps = Omit<ComponentProps<typeof Suspense>, 'fallback'>;

type Props = SuspenseProps &
  ErrorBoundaryProps & {
    rejectedFallback: ComponentProps<typeof ErrorBoundary>['renderFallback'];
    pendingFallback: ComponentProps<typeof Suspense>['fallback'];
  };

interface ResetRef {
  reset?(): void;
}

const AsyncBoundary = forwardRef(function (
  { pendingFallback, rejectedFallback, children, ...errorBoundaryProps }: Props,
  resetRef: Ref<ResetRef>
) {
  const ref = useRef<ErrorBoundary | null>(null);

  useImperativeHandle(resetRef, () => ({
    reset: () => ref.current?.resetErrorBoundary(),
  }));

  return (
    <ErrorBoundary ref={ref} renderFallback={rejectedFallback} {...errorBoundaryProps}>
      <Suspense fallback={pendingFallback}>{children}</Suspense>
    </ErrorBoundary>
  );
});

export default AsyncBoundary;

```

I worried about that If any developer use only @toss/error-boundary, not @toss/async-boundary
they can't use AsyncBoundaryProvider. so they are going to make ErrorBoundaryProvider.

So If AsyncBoundaryProvider's feature will be bigger, I want to suggest to make ErrorBoundaryProvider and wrap it by AsyncBoundaryProvider.
It won't make BREAKING CHANGE too.

**Comment by HoseungJang** [comment]
@raon0211 I think you may know more details of `AsyncBoundary` than me. So could you discuss about this issue with @manudeli ?

But I have a question that `AsyncBoundaryProvider` seems that it does nothing. Because `AsyncBoundary` is not passing `resetKeys` prop to `ErrorBoundary`. I think it should get `resetKey` from `useAsyncBoundaryContext()` and then pass it to `ErrorBoundary`.

**Comment by raon0211** [comment]
Yes, I think @manudeli's opinion is right. Resetting errors should be a responsibility of `@toss/error-boundary`, not `@toss/async-boundary`. Creating `ErrorBoundaryProvider` and making `AsyncBoundaryProvider` to use it seems to be the right way. (Can you refactor this and add a test case? -- we also might do this)

**Comment by raon0211** [comment]
> P.S. I have a question that AsyncBoundaryProvider seems that it does nothing. Because AsyncBoundary is not passing resetKeys prop to ErrorBoundary. I think it should get resetKey from useAsyncBoundaryContext() and then pass it to ErrorBoundary.

Yes the current implementation is wrong. We should fix this. (A test code will show that it's not working)

**Comment by manudeli** [comment]
@raon0211 how about merging @toss/async-boundary with @toss/error-boundary
first time you comment, I thought about refactor seperately.

but If someone use error-boundary, why not async-boundary
they scope same problem. just suspense with errorboundary.

I thought it's why there was no resetboundaryprovider in @toss/boundary.
So I testly made a new package merging two pacakges into [@suspensive/boundary](https://github.com/suspensive/boundary) self.
How about this way?

**Comment by raon0211** [comment]
Sorry for the late reply. 

We might merge the two libraries together, but I don't think that's a must. I prefer smaller libraries to restrict responsibilities and therefore enhance maintainability. Combining two libraries into one is easy, but splitting libraries is hard and cumbersome.

Wouldn't adding `ErrorBoundaryProvider` in `@toss/error-boundary` and making `@toss/async-boundary` use it just be enough? I don't see big reasons to combine two boundary libraries into one.

**Comment by manudeli** [comment]
> Sorry for the late reply.
> 
> We might merge the two libraries together, but I don't think that's a must. I prefer smaller libraries to restrict responsibilities and therefore enhance maintainability. Combining two libraries into one is easy, but splitting libraries is hard and cumbersome.
> 
> Wouldn't adding `ErrorBoundaryProvider` in `@toss/error-boundary` and making `@toss/async-boundary` use it just be enough? I don't see big reasons to combine two boundary libraries into one.

I added resetKeyContext in [this PR](https://github.com/toss/slash/pull/157)
ErrorBoundaryProvider(you named) have only one feature yet resetting ErrorBoundary easily outside of ErrorBoundary fallback or inside of nested components

When any other feature without resetting is necessary for ResetKeyProvider, In my opinion, at that time, we have to naming this component as ErrorBoundaryProvider as new feature of @toss/error-boundary.

but Isn't not necessary any other feature now? so I named this feature as ResetKey like prop of ErrorBoundary's resetKeys.

Any change you want to PR is welcome, I just want to show example that we can see what problem is exactly.

ResetKey require below features
1. preventing bothersome prop drilling
2. reset ErrorBoundary outside of itself
3. reset ErrorBoundary without using resetKeys (prop of ErrorBoundary)

I want to make solution like this
1. useResetKey in ResetKeyProvider
2. use withResetKey wrapping component's props
3. add ErrorBoundary.ResetKey without using resetKeys
